### DISCUSSION TITLE
tui-error not showing errors

### DISCUSSION THREAD

**Discussion by vytautas-pranskunas-** [taiga-family/taiga-ui#9321] [Q&A] [answered]
Hi,

i have this control: 
`
 personalCode: [null, [RxwebValidators.minLength({ value: 3, message: 'test msg' })]]
`

and this htnl
`
 <tui-textfield>
                <input tuiTextfield type="number" formControlName="personalCode" />
            </tui-textfield>
            <tui-error formControlName="personalCode" [error]="[] | tuiFieldError | async" />
`

but message is not beying picked up

**Comment by vytautas-pranskunas-** [comment]
Ok errors not shopwing because of type="number"
not sure why...
but custom error messages are not picked up. But messages persist in form

![Screenshot 2024-10-02 at 21 42 45](https://github.com/user-attachments/assets/c9049184-f4cf-4dc8-b615-3fb67cb7522b)

**Reply by waterplea** [reply]
Please make a Stackblitz example:
https://taiga-ui.dev/stackblitz
Also if you need a number input I strongly recommend `InputNumber` rather than type="number". I know it's legacy still, with no replacement based on new `tui-textfield` yet, but it's a proper numbers input for Taiga UI with formatting, `number` in value, min/max/precision controls etc.

**Reply by hakimio** [reply]
For "Personal identification number" it might be a better idea to use text input with `maskito` mask.
Maybe a similar mask to card mask could be used: https://github.com/taiga-family/taiga-ui/blob/7c43df8cc75c1d37dde82a62499c4cd6d5d6c74c/projects/addon-commerce/constants/mask-card.ts#L4

**Reply by vytautas-pranskunas-** [reply]
I have created stackblitz here: https://stackblitz.com/edit/angular-fdraqy?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.template.html,package.json&hideExplorer=1

**Reply by vytautas-pranskunas-** [reply]
any updates on this? thoughts?

**Reply by waterplea** [reply]
The error is showing for me:
![image](https://github.com/user-attachments/assets/9e0a67ca-7590-4131-9380-2e40b7e6aac5)
I'm not sure what were you expecting. Did you want it to show `"test message"`? How would you expect `tui-error` to know what key to read from the error object? If you add this to providers it would work:
```ts
{
  provide: TUI_VALIDATION_ERRORS,
  useValue: {
    required: ({message}) => message
  }
}
```

**Reply by vytautas-pranskunas-** [reply]
I expect it to read from formControlName

**Reply by hakimio** [reply]
Try out `@ngneat/error-tailor`. You can get rid of `tui-error` boilerplate entirely when using it.
[Make Your Angular Form’s Error Messages Magically Appear](https://netbasal.com/make-your-angular-forms-error-messages-magically-appear-1e32350b7fa5)
[Error Tailor on GitHub](https://github.com/ngneat/error-tailor)

**Reply by vytautas-pranskunas-** [reply]
I understand that i can do it, hut then i do not get point of tuiError. Thats not difficult to show all errors taken from controlvalueaccessor... and it would follow and would be good fit to tui form conponents. So just tell me @waterplea will tui team do outputing all errors from form control to tuiError or should i implement my own error directive or component?

**Reply by hakimio** [reply]
Taiga components are primitive components made by a small team with customizability as first priority and not having all the features out-of-the-box. 
If you need feature rich components which have everything built-in, you should look at commercial products like [Kendo UI for Angular](https://www.telerik.com/kendo-angular-ui) or [DevExtreme Angular](https://js.devexpress.com/Angular/).

**Reply by vytautas-pranskunas-** [reply]
I am aware of alternatives but i do not need complex ui library. However making tuiError to grab errors from formCOntrolName it is primitive functionality because tuiError already reacts to state of the control so display errors would be great "add on" to usability

**Reply by hakimio** [reply]
`tui-error` is a dumb component. All the smarts come from [tuiFieldError pipe](https://github.com/taiga-family/taiga-ui/blob/58f70456de5fb1d8d2d57943948f00109334ee9d/projects/kit/pipes/field-error/field-error-pipe.ts).

**Reply by vytautas-pranskunas-** [reply]
Perfect, then tuiFieldError should grab all errors and displays them when empty array or some param is passed. 
btw i tried

`
[error]="['required'] | tuiFieldError | async"
`

but is till shows just `Value is invalid` instead of actual error message
so is there a way to output all actual error messages rather than `Value is invalid`?

**Reply by hakimio** [reply]
Have you tried providing `TUI_VALIDATION_ERRORS` like @waterplea suggested? [Documentation demo](https://taiga-ui.dev/pipes/field-error#custom-messages).

**Reply by vytautas-pranskunas-** [reply]
i see how it is done now. Thats realy teidly coupled with native validation but if people use something else like @rxweb/reactive-form-validators which has their own validation messaging then people should go other ways. Ok. Thanks

**Reply by waterplea** [reply]
> I expect it to read from formControlName

Inside your `formControlName` there's a dictionary of errors, one of which is:
```
required: {
  message: 'Your text',
  ...
}
```
How do you expect `tui-error`/`tuiFieldError` to figure out that it has to read the `message` key of that arbitrary error object? It shows "Value is invalid" as a default message because it doesn't know how to process arbitrary object that is stored inside the errors dictionary. If it was a string like `required: 'Your text'` it would work

**Reply by waterplea** [reply, marked as answer]
- The idea of `tui-error` is to show arbitrary messages styled as errors, with height/fade transition
- The idea of `tuiFieldError` is to link `tui-error` with form controls
- The idea of `TUI_VALIDATION_ERRORS` is to provide presentation for arbitrary error objects stored inside controls
